.github/workflows/ci-pipeline.yml

```diff
@@ -0,0 +1,45 @@
+name: Continuous Integration
+on:
+  push:
+    paths-ignore:
+      - '**/*.md'
+  pull_request:
+    branches:
+      - main
+    paths-ignore:
+      - '**/*.md'
+
+jobs:
+  build:
+    name: Build and test
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v2
+    - name: Build test docker image
+      run: docker build --target test --tag todo-app:test .
+    - name: Run tests
+      run: docker run todo-app:test
+
+  deploy:
+    name: Push image to Dockerhub
+    runs-on: ubuntu-latest
+    needs: build
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+    steps:
+    - name: Login to Dockerhub
+      uses: docker/login-action@v2
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_TOKEN }}
+    - name: Checkout code
+      uses: actions/checkout@v2
+    - name: Build and push docker image
+      uses: docker/build-push-action@v4
+      with:
+        context: .
+        push: true
+        target: production
+        tags: ${{ secrets.DOCKERHUB_USERNAME }}/todo-app:latest,${{ secrets.DOCKERHUB_USERNAME }}/todo-app:${{ github.sha }}
+    - name: Deploy to Azure
+      run: curl -dH -X POST '${{ secrets.AZURE_WEBHOOK }}'
```

Comment:
One nice addition here can be adding the [--fail](https://curl.se/docs/manpage.html#-f) flag, as otherwise curl reports success as long as a successful connection was made, even if the response reports a failure (e.g. a status code > 400) which can lead to confusing pipeline results